### PR TITLE
Some refactoring on PidsLimit

### PR DESCRIPTION
**Pull request by thaJeztah**
Some follow-ups to https://github.com/moby/moby/pull/38791 ~(first commit is https://github.com/moby/moby/pull/38791)~

### Normalize values for pids-limit

- Don't set `PidsLimit` when creating a container and no limit was set (or the limit was set to "unlimited")
- Don't set `PidsLimit` if the host does not have pids-limit support (previously "unlimited" was set).
- Do not generate a warning if the host does not have pids-limit support, but pids-limit was set to unlimited (having no limit set, or the limit set to "unlimited" is equivalent, so no warning is nescessary in that case).
- When updating a container, convert `0`, and `-1` to "unlimited" (`0`).

### Update TestUpdatePidsLimit to be more atomic

Create a new container for each subtest, so that individual subtests are self-contained, and there's no need to execute them in the exact order, or resetting the container in between.

This makes the test slower (6.54s vs 3.43s), but reduced the difference by using `network=host`, which made a substantial difference (without `network=host`, the test took more than twice as long: 13.96s).

### CODE REVIEW

File: integration/internal/container/ops.go [hunk 1]
```diff
@@ -142,3 +142,13 @@ func WithAutoRemove(c *TestContainerConfig) {
 	}
 	c.HostConfig.AutoRemove = true
 }
+
+// WithPidsLimit sets the container's "pids-limit
+func WithPidsLimit(limit *int64) func(*TestContainerConfig) {
```
Comment (thaJeztah) [inline integration/internal/container/ops.go]: Was in doubt if I should change this to a non-pointer to make it easier to use (and if no limit is needed, don't pass `WithPidsLimit()`)

Comment (cpuguy83) [reply integration/internal/container/ops.go]: I can see both ways here. In the end I think making a pointer is simpler than dealing with filtering options... e.g.


```go
opts := []ConfigOpts{WithFoo,WithBar}
if hasLimit {
  opts = append(opts, WithPidsLimit(limit))
}
container.Run(..., opts...)
```

File: integration/container/update_linux_test.go [hunk 2]
```diff
@@ -114,42 +113,35 @@ func TestUpdatePidsLimit(t *testing.T) {
...
 
-		var before types.ContainerJSON
-		if test.update == nil {
-			var err error
-			before, err = c.ContainerInspect(ctx, cID)
-			assert.NilError(t, err)
-		}
-
 		t.Run(test.desc, func(t *testing.T) {
+			// Using "network=host" to speed up creation (13.96s vs 6.54s)
```
Comment (MHBauer) [inline integration/container/update_linux_test.go]: seems like a valid choice. we're not doing anything with the network here.

File: api/server/router/container/container_routes.go [hunk 3]
```diff
@@ -481,6 +488,14 @@ func (s *containerRouter) postContainersCreate(ctx context.Context, w http.Respo
 		hostConfig.Capabilities = nil
 	}
 
+	if hostConfig != nil && hostConfig.PidsLimit != nil && *hostConfig.PidsLimit <= 0 {
+		// Don't set a limit if either no limit was specified, or "unlimited" was
+		// explicitly set.
+		// Both `0` and `-1` are accepted as "unlimited", and historically any
+		// negative value was accepted, so treat those as "unlimited" as well.
+		hostConfig.PidsLimit = nil
```
Comment (MHBauer) [inline api/server/router/container/container_routes.go]: good comment

File: daemon/daemon_unix.go [hunk 4]
```diff
@@ -119,16 +119,16 @@ func getMemoryResources(config containertypes.Resources) *specs.LinuxMemory {
...
-			limit.Limit = -1
-		}
+	if config.PidsLimit == nil {
+		return nil
 	}
-	return limit
+	if *config.PidsLimit <= 0 {
+		// docker API allows 0 and negative values to unset this to be consistent
+		// with default values. When updating values, runc requires -1 to unset
+		// the previous limit.
```
Comment (MHBauer) [inline daemon/daemon_unix.go]: nice cleanup of the comment.

**Review by MHBauer — COMMENTED** [review]
needs rebase on #38791 I think?

**Comment by thaJeztah** [comment]
rebased; PTAL

File: integration/internal/container/ops.go [hunk 5]
```diff
@@ -137,6 +137,16 @@ func WithAutoRemove(c *TestContainerConfig) {
 	c.HostConfig.AutoRemove = true
 }
 
+// WithPidsLimit sets the container's "pids-limit
+func WithPidsLimit(limit *int64) func(*TestContainerConfig) {
+	return func(c *TestContainerConfig) {
+		if c.HostConfig == nil {
```
Comment (cpuguy83) [inline integration/internal/container/ops.go]: nit: I don't think this is needed at this point.

**Review by cpuguy83 — APPROVED** [review]
Oy, the more I look at this the more I'm worried about the change we made to support updates for pids limit.

This change LGTM, though.

File: integration/internal/container/ops.go (hunk 5 above)
Comment (thaJeztah) [reply integration/internal/container/ops.go]: hm.. possibly; I might have run into a situation where it was nil, but perhaps that was when I was working on the test. Don't think it hurts to have it :sweat_smile:

**Comment by thaJeztah** [comment]
> Oy, the more I look at this the more I'm worried about the change we made to support updates for pids limit.

I was trying to make the flow more consistent (some attempts in https://github.com/moby/moby/pull/38676). I think one of the problems we have currently is that in many cases, the (`Host-`)`Config` we store on disk is reflecting the JSON we sent over the API. Perhaps this was the design, but it means that API conventions trickle down all the way into the daemon, so when converting that config to an OCI spec, we have to keep the API conventions ("over the wire format") into account.

I've been following some of the code-paths, and validation of values is interweaved throughout the codebase; here's some notes I was taking;

```
-> postContainersCreate
  -> containerRouter.decoder.DecodeConfig
    -> runconfig.decodeContainerConfig
      -> ContainerConfigWrapper.getHostConfig (modifies config for backward-compatibility)
      -> validateNetMode
        -> validateNetContainerMode
      -> validateIsolation
      -> validateQoS
      -> validateResources (checks if kernel supports requested options)
      -> validatePrivileged (platform-specific validation; error on Windows for privileged)
      -> validateReadonlyRootfs (platform-specific validation; error on Windows as it's not supported)
  -> do some manipulation of values related to API versions
  -> backend.ContainerCreate
    -> daemon.containerCreate
      -> daemon.adaptContainerSettings (updates passed configuration, also does some validation)
      -> daemon.validateContainerConfig
        -> translateWorkingDir (modifies working-dir based on platform)
      -> daemon.validateContainerHostConfig
        -> validateHostConfig
        -> verifyPlatformContainerSettings
      -> daemon.verifyNetworkingConfig
```

I _think_ what would be nice if we would handle the API "conversion" in the API code (where possible), so that the config as stored on disk could more easily be converted to an OCI-spec without having to handle those conversions (again), and an "update" would be more of a "reconciliation". Validation might become easier then as well.

Probably that would be problematic for upgrade scenarios (configs created by older versions of the daemon), and there's likely many things I didn't consider.


Also; I keep looking at our tests, and thinking that the client we use in our tests has a much nicer interface than the API client (with the functional arguments) (disclaimer: I've become a bit of a fan of those :sweat_smile:)

**Review by MHBauer — APPROVED** [review]
I thought I had decoupled the API, but it looks like it's back to sharing structs.

LGTM

**Pull request merged**